feat: add user mapping and model migrations tables

kian99 opened the pull request:
## Description
This PR adds 2 new tables called `model_migrations` and `user_mappings` which will be used when migrating models to JIMM. In full this PR adds,
- 1 new SQL schema migration.
- The Go representations for the new structs in `internal/dbmodel`
- CRUD methods in `internal/db`, not all CRUD methods have been added, only those that make sense.

Fixes [JUJU-7950](https://warthogs.atlassian.net/browse/JUJU-7950)

## Engineering checklist

- [ ] Documentation updated
- [x] Covered by unit tests
- [ ] Covered by integration tests

## Test instructions
<!-- *(optional)* Describe any non-standard test instructions and configuration settings. Delete this section if not applicable. -->


[JUJU-7950]: https://warthogs.atlassian.net/browse/JUJU-7950?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kian99 committed:
feat: add user mapping and model migrations tables

SimoneDutto commented on internal/dbmodel/usermapping.go:
```diff
@@ -0,0 +1,27 @@
+	// LocalUser is the local user that this mapping applies to.
+	LocalUser string
+
+	// ExtneralUserName is the external user that this local user maps to.
```
```suggestion
	// ExternalUserName is the external user that this local user maps to.
```

SimoneDutto commented on internal/db/modelmigration.go:
```diff
@@ -0,0 +1,79 @@
+)
+
+// AddModelMigration stores information about an incoming model migration.
+//   - returns an error with code errors.CodeAlreadyExists if
```
i like this type of comment.

SimoneDutto commented on internal/dbmodel/modelmigration.go:
```diff
@@ -0,0 +1,25 @@
+	TargetController   Controller
+
+	// UserMapping holds a mapping of local users to external users.
+	UserMapping JSON
```
Why do we hold this JSON reference?
At the end of the day we use the JSON to create the user mappings in our db, so we can get rid of the original JSON mapping right?

kian99 replied:
Correct. This row for model-migration will be deleted when the migration completes/aborts. If it succeeds we will take the user-mapping JSON and apply it to the user-mapping table.

SimoneDutto commented on internal/dbmodel/usermapping.go:
```diff
@@ -0,0 +1,27 @@
+	// Note this doesn't use the standard gorm.Model to avoid soft-deletes.
+	ID        uint `gorm:"primarykey"`
+	CreatedAt time.Time
+	UpdatedAt time.Time
```
There isn't any method to update the `UserMapping`.
If we'll create that since we are not using `gorm.Model` remember to manually handle the UpdateAt.

SimoneDutto commented on internal/dbmodel/sql/postgres/024_add_model_migration_tables.up.sql:
```diff
@@ -0,0 +1,27 @@
+    user_mapping JSONB NOT NULL
+);
+
+CREATE INDEX idx_model_migrations_model_uuid ON model_migrations (model_uuid);
```
I don't think you need this one, because you already have `model_uuid TEXT NOT NULL UNIQUE`, unless you are doing this to improve queries performance, which I don't think would ever be a problem on this table.

kian99 replied:
True, and also 
>[PostgreSQL automatically creates a unique index when a unique constraint or primary key is defined for a table.](https://www.postgresql.org/docs/current/indexes-unique.html#:~:text=PostgreSQL%20automatically%20creates%20a%20unique,mechanism%20that%20enforces%20the%20constraint.)

SimoneDutto reviewed (commented):
looking good, some comments

SimoneDutto commented on internal/dbmodel/sql/postgres/024_add_model_migration_tables.up.sql:
```diff
@@ -0,0 +1,27 @@
+    CONSTRAINT unique_user_mappings_key UNIQUE(model_uuid, local_user)
+);
+
+CREATE INDEX idx_model_user_mappings_local_user ON user_mappings (model_uuid, local_user);
```
I don't think you need this either. It would basically index each row.

Maybe you are looking for CONSTRAINT

"The preferred way to add a unique constraint to a table is ALTER TABLE ... ADD CONSTRAINT" 
https://www.postgresql.org/docs/9.4/indexes-unique.html

kian99 replied:
When we query for a user mapping, we will always query by the model-uuid and the local-user (these are the 2 pieces of information in the macaroon, alongside the offer-uuid). That's why I created the multicolumn index.

ale8k replied:
The table would always be small right? So wouldn't the db sequentially scanning be quicker than an index? I think remove this index

ale8k replied:
See: https://funivan.com/post/index-scan-vs-sequential-scan/ for more

kian99 replied:
It doesn't matter much either way as you say the table will likely be small. But the expected access pattern is such that we are always querying by model-uuid and local user. In any case, I'll just make it an index on the model UUID.

ale8k replied:
What I'm saying is no index is probably best for this

kian99 replied:
Would definitely require some actual measurements to know for sure, but usually an index helps so I'll leave this as is.

kian99 committed:
chore: remove unnecessary index

ale8k commented on internal/dbmodel/modelmigration.go:
```diff
@@ -0,0 +1,25 @@
+	// ModelUUID is the UUID of the incoming model.
+	ModelUUID sql.NullString
+
+	// TargetControllerID is the target controller for the model undegroing migration.
```
s/undegroing/undergoing

ale8k commented on internal/dbmodel/modelmigration.go:
```diff
@@ -0,0 +1,25 @@
+)
+
+// ModelMigration holds the information for a pending model migration.
+type ModelMigration struct {
```
This needs a different name, this isn't model migrations, this is "ModelMigrationTargetAndInitialUserMappingsToStoreOnCompletion" (Just describing it aha, but a better name is my point) and I also think you should godoc its explicit purpose (specifically UserMapping) and how it'll be used. We'll lose context in time with this, so I'd prefer us be kind to ourselves

kian99 replied:
Here are some name options `ActiveModelMigrations` or `ModelMIgrationJobs` or `OngoingModelMigrations`. But I'm not sure about including details about the fields in the table in the name.

kian99 replied:
I've updated the godoc though.

ale8k replied:
Whatever you think if best, I just didn't think calling it ModelMigration was right. Happy with any of them honestly.

alesstimec commented on internal/db/modelmigration.go:
```diff
@@ -0,0 +1,79 @@
+
+	db := d.DB.WithContext(ctx)
+
+	if err := db.Create(modelMigration).Error; err != nil {
```
should we verify the UserMapping field to make sure it contains something we expect and not a random json blob?

kian99 replied:
Yes but not at the DB layer

ale8k commented on internal/db/modelmigration.go:
```diff
@@ -0,0 +1,79 @@
+}
+
+// GetModelMigration returns model migration information based on the model UUID.
+func (d *Database) GetModelMigration(ctx context.Context, modelMigration *dbmodel.ModelMigration) (err error) {
```
I know we usually pass the dbmodel obj, but would love to see us just return a dbmodel.ModelMigration and the func take a uuid. Really hate this style and want us to move away from it honestly.

kian99 replied:
I'd prefer to just follow the pattern, so leaving this as is.

ale8k commented on internal/dbmodel/usermapping_test.go:
```diff
@@ -0,0 +1,44 @@
+// TestUserMapping_UniqueConstraint tests that the unique constraint on
+// UserMapping is enforced correctly ensuring two rows with the same
+// ModelUUID and LocalUser cannot be created.
+func TestUserMapping_UniqueConstraint(t *testing.T) {
```
Interesting we using _ here, looks like AI added this because we don't usually?

kian99 replied:
Haha I was wondering if anyone would notice this. I actually added it because I saw AI adding somewhere else in Juju and I liked it. Splits the thing I'm testing with the specifics.

ale8k replied:
I really like it (Imo) but we discussed it once witn Juju and they hated it, and I think I tried to push for it for us and rightfully so, @alesstimec wasn't keen because it's just not how we do things and stood out like a sore thumb. having said that, I'm 1000% for continuing like this if the guys are on board?

kian99 replied:
I don't mind, but since it's different I've just opted to make it consistent now.

ale8k replied:
Aww, we can discuss it another time though I am really in favour of it.

alesstimec commented on internal/db/modelmigration_test.go:
```diff
@@ -0,0 +1,155 @@
+	c.Assert(dbMigration.UserMapping, qt.DeepEquals, migration.UserMapping)
+}
+
+func (s *dbSuite) TestAddModelMigrationAlreadyExists(c *qt.C) {
```
this case could well be covered in the test above, which would avoid duplicating the same setup..

kian99 replied:
Fair, I'll remove the duplication.

ale8k commented on internal/dbmodel/sql/postgres/024_add_model_migration_tables.up.sql:
```diff
@@ -0,0 +1,25 @@
+-- The model_migration table contains information
+-- about ongoing migrations and their target controller.
+-- The UserMapping table is used to map local users
+-- from migrated models to their external counterparts.
```
Where is the version bump? Or have you purposefully left it until a later PR?

kian99 replied:
Which version bump? Our migrations run up to the latest, we don't need to specify the number in a file somewhere anymore.

alesstimec commented on internal/db/modelmigration_test.go:
```diff
@@ -0,0 +1,155 @@
+	}
+	c.Assert(s.Database.DB.Create(&migration).Error, qt.IsNil)
+
+	err = s.Database.DeleteModelMigration(context.Background(), &migration)
```
perhaps also try to delete model migration that does not exist?

kian99 replied:
Added it, but I see it doesn't return an error when deleting an item that doesn't exist. So, do you still want it?

ale8k commented on internal/dbmodel/modelmigration_test.go:
```diff
@@ -0,0 +1,40 @@
+		TargetControllerID: ctl.ID,
+		UserMapping:        dbmodel.JSON([]byte(`{"new-local": "new-external"}`)),
+	}
+	c.Assert(db.Create(&m2).Error, qt.ErrorMatches, ".*duplicate key value violates unique constraint \"model_migrations_model_uuid_key\".*")
```
@alesstimec often prefers we store result of errors and then c.Assert(err...), shouldn't we keep to doing that?

kian99 replied:
Left this as is, Ales didn't comment on it.

alesstimec commented on internal/db/usermapping_test.go:
```diff
@@ -0,0 +1,105 @@
+		ExternalUserName: identityName,
+	}
+
+	err = s.Database.AddUserMapping(ctx, userMapping)
```
try adding it twice and assert that you get already exists error

alesstimec commented on internal/db/usermapping_test.go:
```diff
@@ -0,0 +1,105 @@
+	c.Assert(s.Database.AddUserMapping(ctx, userMapping), qt.IsNil)
+
+	// Delete
+	err = s.Database.DeleteUserMapping(ctx, userMapping)
```
try deleting user mapping that does not exist

ale8k commented on internal/db/usermapping.go:
```diff
@@ -0,0 +1,77 @@
+	case userMapping.ModelUUID.Valid && userMapping.LocalUser != "":
+		db = db.Where("model_uuid = ? AND local_user = ?", userMapping.ModelUUID.String, userMapping.LocalUser)
+	default:
+		return errors.E(op, "missing model uuid and/or local user", errors.CodeBadRequest)
```
Be nicer to say exactly what is missing with something like:
```go
switch {
case !userMapping.ModelUUID.Valid && userMapping.LocalUser == "":
	return errors.E(op, "missing both model UUID and local user", errors.CodeBadRequest)
case !userMapping.ModelUUID.Valid:
	return errors.E(op, "missing model UUID", errors.CodeBadRequest)
case userMapping.LocalUser == "":
	return errors.E(op, "missing local user", errors.CodeBadRequest)
default:
	if err := db.First(&userMapping).Error; err != nil {
		err = dbError(err)
		if errors.ErrorCode(err) == errors.CodeNotFound {
			return errors.E(op, err, "user mapping not found")
		}
		return errors.E(op, dbError(err))
	}
}
```

kian99 replied:
Will do

ale8k commented on internal/db/modelmigration_test.go:
```diff
@@ -0,0 +1,155 @@
+	"github.com/canonical/jimm/v3/internal/errors"
+)
+
+func TestAddModelMigrationUnconfiguredDatabase(t *testing.T) {
```
We add these all over the place, I don't think we should be duplicating it constantly... Maybe just delete this?

kian99 replied:
Sure

kian99 committed:
chore: remove test duplication

kian99 committed:
chore: remove "unconfigured db" test

SimoneDutto reviewed (approved):
Lgtm

kian99 committed:
Merge branch 'v3' into model-migrations

ale8k reviewed (approved):
lgtm

kian99 committed:
chore: change ModelMigration to IncomingModelMigration

Changed the name of the table to better reflect that these are model being migrated into JIMM rather than away.
